### PR TITLE
Add introduction to TPM concepts

### PR DESCRIPTION
**Pull request by nicowilliams**
This probably requires careful review before merging.

My goal is to have an intro suitable for technical people who are lay when it comes to TPMs.  This is the introduction I would have liked to have had a few months ago, along with the attestation tutorial.

### CODE REVIEW

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md]: Leave this as "to simplify things...".     For a newbie, WTF is algorithm agility? :-)

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md:8]: This is an attempt at a simple explanation --
much simpler than reading hundreds of pages of TCG  documents and this will at least get you familiar with the concepts.

Comment (iolivergithub) [inline Intro/README.md:8]: There's stuff here that you are free to take and incorporate : https://github.com/nokia/TPMCourse/tree/master/docs
Check the license though :-)

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md]: You could reduce this down to the main concepts - this is a huge list:

- TPM Objects and Hierarchies
- Keys
- NVRAM
- PCRs and Measurements
- Attestation
- Advanced Stuff
- --- SEssions, Tickets, etc

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md]: We'll assume reader familiarity with the *basics* of cryptographic keys and how encryption, decryption, signing and verification work with asymmetric and symmetric keys, eg: RSA.

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md]: I think this needs to be restructued - I'm already lost here.  To introduce authorisations we need to introduce hierarchies first and before that what a TPM has inside, ie:  Objects and Hierarchies..   Objects can be keys, NVRAM slots, PCRs etc,  Hierarchies - we have 4 of though .

File: Intro/README.md
Comment (iolivergithub) [inline Intro/README.md]: They don't need to be, but the spec says so for security reasons.

**Review by iolivergithub — COMMENTED** [review]
Not finished, but I think restructuring is required. LOTS of stuff being introduced here.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: It is a huge list, and I like that: it conveys how complex the topic is.  I don't mean to cover every item though.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md:8]: Will do.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I kinda disagree though.  I think EA policies are a perfect example of the power of hash extension, and extremely clever.  EA policies aren't all that interesting as an introductory matter, but their cleverness is, and, IMO, it is authorization in all its flavors that makes TPMs so useful.  I want to give the reader a flavor for the flexibility and power of TPMs, but I want to do it in just a few pages, not hundreds.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Software vTPMs aren't slow.  The issue is lack of resources -- HW TPMs have to be _cheap_.

Comment (nicowilliams) [reply Intro/README.md]: If the spec says they have to be slow, I've missed it.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md:8]: Linking to it should be fine!

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I've read "A Practical Guide to TPM 2.0", and I highly, highly recommend that book.  It worked for me.  

Because all these TPM topics are so interrelated you get the old `B before A` pattern.  Want to learn quantum mechanics?  You're going to have to learn... so much other stuff before you start on that.  But for many students that approach is just discouraging -- for them "let's study just enough of B to get started on A, then go back to B later" works very well.

In the case of TPMs there's just... so much of of the `B before A` pattern, and even some circularity...  You really see a fair bit of this in "A Practical Guide to TPM 2.0", where many chapters cover some related topic lightly and then later chapters cover that ancillary topic in depth.

I don't mind the `B- before A then A then B+ then A+` pattern, but I need a more accessible document for people who mostly don't need to become experts in TPMs, and hopefully such a document can help _some_ of its audience get started on becoming experts.

One has to start somewhere.  Where?  I think hash extension is the obvious starting place because it is core to PCRs and policies.  Also cryptographic object naming: because it too is central to authorization!  And authorization is, IMO, _the_ essential feature of TPMs, so it is my goal to get to it, and I want to get to it very quickly because I don't want to bore and drive away the lay reader.

Obviously explaining EA policies in all their detail requires first explaining everything else.  Yet since policies are so interesting, leaving them for last is a bit disappointing.  My approach right now is: explain a little bit of each concept, then explain policies, then stop.

My take is that key hierarchies are not as essential to understanding policies as you seem to think.  Maybe... maybe that's wrong.  I don't know.  Given my goal here I need to hear from reviewers who literally know nothing about TPMs.  I also want review by TPM experts, both as to errors in my text, and as to opinion on how to structure this, but while I will consider such opinions and maybe accept them, I might also reject some opinions.

Eventually I -or someone else- might write in-depth coverage of each topic, but still having this very light umbrella treatment should help approach each topic's in-depth coverage.

File: Intro/README.md
Comment (jheiss) [reply Intro/README.md]: The list is a bit overwhelming to me. At least on the first reading my eyes just skip over it since "Some core concepts in the world of TPMs" doesn't convey that the list is terribly important, and I know I'll never remember 90% of it anyway.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: What does "legitimate hardware TPMs that otherwise could not be secured" mean?

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: "critical dependence on TPMs performing authorization as specified": Is this any different than "we expect computers to compute properly"? I'm not sure why I wouldn't expect the TPM to behave correctly. Are you getting at whether the TPM is trustworthy? And whether that trustworthiness is different than the rest of the computer hardware? If so you might need to explain that a bit more.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: I'm not sure in an intro doc this long list of things the TPM can't do is necessary, particularly as some of them are pretty esoteric.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: "mure" typo

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md:91]: That's a big Important Notes section between "Let's start with hash extension" and actually talking with hash extension. Consider restructuring to bring those two closer together.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: Maybe "the system has only executed trusted code"? Just "only executed trusted code" isn't super clear to me

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: "eventlot" typo

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: The parenthetical is perhaps not necessary in an intro doc.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: Maybe instead of "symmetric key or asymmetric private key" just say "private key" here? It took me a few tries to parse this sentence.

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: Is the exporting of keys, or at least this parent business important enough for this doc? It's confusing, and doesn't seem like a thing that would be commonly used?

File: Intro/README.md
Comment (jheiss) [inline Intro/README.md]: Does this mean that the key hierarchy described earlier gets reset on every boot by default? That seems surprising if so, and perhaps worthy of a bit more explanation.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Ah, I've something very specific in mind here, which is restricted keys, specifically `TPM2_ActivateCredential()` and the other side of it, `TPM2_MakeCredential()`, and basically you only want to send a "credential" to a TPM you know is legitimate.  Except of course that the TPM might be a virtual TPM, implemented in software in a hypervisor, and then what?  Well, the idea is that you authenticate the TPM's EKpub/EKcert before you call `TPM2_MakeCredential()` to send it a credential.

OK, so maybe this is too soon to hint at that and I should just remove that text.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Perhaps, yeah.  I guess what I want to say is: TPMs only do very basic things because they are very limited.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I'll think about re-phrasing this.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Hmm, to me "private key" implies asymmetric crypto, and "secret key" implies symmetric crypto.  I'll re-phrase.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Actually, yes, I think so.  Perhaps quite a bit more should be said here about key wrapping.  The important thing is to note that one can have keys that are guaranteed not to leave the TPM in the clear nor exportable to other TPMs or systems in any way.  A key that is fixed to a TPM can still be saved, but only reloaded into the same TPM, though you can also have keys that can't be saved either.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: There are non-volatile keys and volatile keys, and the latter are "gone" on restart, but if you can save them then you can reload them post-restart.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: `s/that otherwise could not be secured//`

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: Maybe instead we could mention that the TPM is a passive device?

Including that a TPM only responds to request from the Host rather than act on its own.

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: Follow-up from our TPM.dev weekly call this week: @kgoldman checked in the spec and said that a restricted key is a TPM key that can be used only on TPM-generated data. For example, Attestation Keys can only sign TPM-generated data. However, the restricted key does not have a fixedTPM attribute (which is what I thought).

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: I have also missed mentioning of that in the spec. In any case, the TPM is designed for security, therefore a trade-off with performance is made. In contrast, modern x86 CPUs have speculative branching and execution to trade performance for security (the meltdown saga) ;)

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: TPM 2.0 Authorization and Extended Authorization(EA) is one of the most complex topics. 

I think I would read through the tutorial once again after few days, so I have fresh eyes and then will give feedback on this paragraph.

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: Good addition @iolivergithub

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md:8]: @nicowilliams I also think linking to external sources is totally fine. 

Nokia's TPM Course is a great resource. Cheers @iolivergithub

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I've removed that whole section.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Removed.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md:91]: Fixed.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I've done some restructuring.

Comment (nicowilliams) [reply Intro/README.md]: > TPM 2.0 Authorization and Extended Authorization(EA) is one of the most complex topics.

But I don't cover it in detail.  And I think it's especially illustrative of the brilliance of using digest values to identify things that don't fit in a TPM's resources.  PCRs and policies both use hash extension in this way.  One needs an externally-maintained eventlog to reconstruct how a PCR got to some value.  One needs to know a policy in order to send it to the TPM for evaluation, then at the end the TPM checks that its digest _is_ the policy associated with whatever resource is being accessed.

I might further restructure the intro to make this particular idea more prominent but cover policies in less detail.

Comment (nicowilliams) [reply Intro/README.md]: Note that, as complex as EA policies are, I don't have that much text on the topic -- ~100 lines (including empty lines) out of ~500 lines of text altogether (also including empty lines).

Comment (kgold2) [reply Intro/README.md]: ".. which may be the only trivial concept in the world of TPMs!"  Too much opinion for my taste.  There's a lot of "this is really hard" throughout.

File: Intro/README.md
Comment (kgold2) [reply Intro/README.md]: " Hardware TPMs are _slow_."  Another opinion.  Slow compared to what.  It would be more useful to just give performance numbers and let the reader decide.

File: Intro/README.md
Comment (kgold2) [reply Intro/README.md]: > 
> Follow-up from our TPM.dev weekly call this week: @kgoldman checked in the spec and said that a restricted key is a TPM key that can be used only on TPM-generated data. 

It can sign any message that the TPM has hashed.  An externally generated message cannot start with the magic number.

File: Intro/README.md
Comment (kgold2) [reply Intro/README.md]: Inaccurate.  A primary key can be whatever you want it to be.  It doesn't even have to be a key.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Compared to the CPU.

Comment (nicowilliams) [reply Intro/README.md]: I didn't think there would be any controversy about "HW TPMs are slow"...

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: Ok.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I pared it down a bit.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: @kgold2 hmm, but in the EK hierarchy the primary key has to be a decrypt-only key, no?

Comment (kgold2) [reply Intro/README.md]: > 
> 
> @kgold2 hmm, but in the EK hierarchy the primary key has to be a decrypt-only key, no?

There is no EK hierarchy.
There is an endorsement hierarchy, and I can create any primary object with any attributes I like.

I encourage you to download the IBM TSS.  You can test these assertions in seconds.

File: Intro/README.md
Comment (kgold2) [reply Intro/README.md]: That's inaccurate.  It depends on the hierarchy.
The TCG uses the terms transient and persistent object.  I encourage you to use the TCG terms.  Using different terms leads to inaccuracies.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: I've removed that section altogether, @kgold2, but https://github.com/wolfSSL/wolfTPM does have benchmarks, and... HW TPMs do look pretty slow to me by comparison to CPUs and, therefore presumably TPM simulators running on CPUs.

File: Intro/README.md
Comment (tomoveu) [reply Intro/README.md]: I am also in favour of the terms transient and persistent because of two reasons

1. They are understandable as is
- transient - something temporary
- persistent - something that remains
2. Both terms seem to already have popularity among people who have tried or are using TPM

@kgold2 I would not mind having an explanation in the TPM intro about where exactly are persistent keys stored. AFAIK, it is in a segment of the non-volatile TPM memory. Perhaps, this is why @nicowilliams went with the terms volatile and non-volatile keys.

**Comment by tomoveu** [comment]
I must admit, this tutorial is going great.

Admirations to everyone contributing @jheiss, @kgold2, @iolivergithub , I will add you to the list of contributors on the main README.md. 

OT: I think Ken and Ian already have a TPM.dev T-shirt, of the first ten ever made :) I just ordered a second batch of 10 T-shirts. Jason, could you please drop a message at dimi@tpm.dev, so I could send you a TPM.dev T-shirt as a thank you?

Kudos @nicowilliams for taking this shot!

File: Intro/README.md
Comment (kgold2) [reply Intro/README.md]: > I would not mind having an explanation in the TPM intro about where exactly are persistent keys stored. AFAIK, it is in a segment of the non-volatile TPM memory.

They are in the TPM's non-volatile memory.  I don't understand 'a segment' in this context.

**Comment by nicowilliams** [comment]
I rebased, but otherwise made no additional changes since a few days ago.

**Comment by nicowilliams** [comment]
I think I've addressed most comments.

File: Intro/README.md
Comment (nicowilliams) [reply Intro/README.md]: > Perhaps, this is why @nicowilliams went with the terms volatile and non-volatile keys.

I'm still fairly new to this space, so I've not fully internalized the terminology, that's all.

> I don't understand 'a segment' in this context.

I don't think @tomoveu meant anything like "segment" in the sense of, say, segmented memory architectures.  I suspect what @tomoveu had in mind is that NV allocated for objects other than NV indexes is... does not get an NV index, so it's "separate" in some sense, though it can be a single pool of NV that is shared for all persistent objects _including_ NV indexes.

Comment (kgold2) [reply Intro/README.md]: > all persistent objects _including_ NV indexes.

NV Indexes are not persistent objects.  (NV Indexes are not objects).

I'm commenting about this because a tutorial that redefines TCG standard terms will lead to confusion.

The TPM specification has a glossary at the beginning of Part 1.

**Comment by tomoveu** [comment]
@nicowilliams I will make one final review over the weekend and merge.

ps: I like the links you added in the README.md

**Review by tomoveu — APPROVED** [review]
The tutorial became great.

Good new additions about the storage key and hierarchy's seed.

**Pull request merged**
